core/src/main/scala/kafka/network/SocketServer.scala

```diff
@@ -677,12 +681,26 @@ private[kafka] class Processor(val id: Int,
         openOrClosingChannel(receive.source) match {
           case Some(channel) =>
             val header = RequestHeader.parse(receive.payload)
-            val context = new RequestContext(header, receive.source, channel.socketAddress,
+            val connectionId = receive.source
+            val context = new RequestContext(header, connectionId, channel.socketAddress,
               channel.principal, listenerName, securityProtocol)
+
+            // This callback is used for quota managers to throttle the channel on quota violation. More specifically,
+            // this will be called (1) when throttling starts, to tell the channel that it cannot be unmuted until
+            // the throttling is done, and (2) when throttling is done, to tell the channel that it can be unmuted as
+            // long as the response has already been sent out to the client.
+            def channelThrottlingCallback(throttledChannelEvent: ThrottledChannelEvent) = {
+              throttledChannelEvent match {
+                case StartThrottling => incrementChannelUnmuteRefCount(connectionId)
+                case EndThrottling => tryUnmuteChannel(connectionId)
```

Comment:
I haven't gone through the PR in detail, but I think this callback is invoked on a different thread when throttle time expires. `Selector` is not thread-safe and we expect methods on the selector to be invoked from a single thread.